### PR TITLE
Add namespace annotation inheritance to pods

### LINKED ISSUE #5977: Support more config annotations on namespaces
## Feature Request

The proxy injector does not ensure pods inherit `config.linkerd.io` annotations from their namespace if they do not have one specified themselves. The exception for this is the `opaque-ports` annotation. #5941 added this inheritance, but also added additional handling for properly displaying when workloads were annotated or injected.

Those changes should be extended to handle not only the `opaque-ports` annotation, but all `config.linkerd.io` annotations.

As more of these annotations are added, it will be helpful to be able to annotate a namespace instead of editing all `spec.templates`s within the namespace.

What comes to mind when writing this is the `proxy-await` annotation. In order for it to be supported in the Viz and Jaeger extensions, every deployment within these extensions needs the annotation added. It would be convenient if we could just add `config.linkerd.io/proxy-await: "enabled"` to their namespaces.

### PR DESCRIPTION
**Pull request by mateiidavid**
Closes #5977  

## What

This changes adds support for namespace configuration annotation inheritance for pods. Any annotations (e.g `config.linkerd.io/skip-outbound-ports` or `config.linkerd.io/proxy-await`) that are applied against a namespace will now also be applied to pods running in that namespace by the _proxy-injector_. 

* Pods do not inherit annotations from their namespaces; the exception to this is `opaque-ports` introduced in #5941. This expands on the work by allowing all config annotations to be inherited.
* Main advantage here is that instead of applying annotations on a workload-by-workload basis we can just apply them against the namespace and it will be mirrored on all pods within the namespace.
* Through this change the controller can also check the proxy's configuration directly from the pod's meta rather than from env variables.

## How

Change is pretty straightforward. We want to make sure that before we apply a JSON patch we first copy all of the namespace annotations to the pod. The logic that was in place takes care of applying the patch.

* One obvious constraint is that we want only want valid configuration annotations to be applied. To be a "valid" configuration it has to exist and it has to be prefixed with `config.linkerd.io` -- the easiest way to do this is to go through all of the available proxy configuration options and check whether any of the options are included in the namespace's annotations (done in `GetNsConfigKeys()` where we fetch all annotation keys from the namespace).
* A consideration I had with this change is whether to add `opaque-ports` as part of all of the config keys; opaque ports is a bit different though since it can be applied on a pod as well as a service -- through this change we only want to apply config annotations to pods. I chose to keep the two separate.
* Added a unit test that checks if a pod inherits config annotations from its namespace; this also includes an invalid annotation which doesn't show up in the "expected" patch to test we validate configuration correctly.

### Tests
---

I injected emojivoto and added an annotation to its namespace:

```
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    config.linkerd.io/opaque-ports: "34567"
    config.linkerd.io/proxy-log-level: debug
    config.linkerd.io/skip-outbound-ports: "44556"
    linkerd.io/inject: enabled
```

The deployment specs do not have any additional annotations as part of the pod template metadata. I first tested if the above annotations would be inherited with the current edge release (I expected opaque ports to be).

**Before changes**:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    config.linkerd.io/opaque-ports: "34567"
    linkerd.io/created-by: linkerd/proxy-injector edge-21.4.1
    linkerd.io/identity-mode: default
    linkerd.io/inject: enabled
    linkerd.io/proxy-version: edge-21.4.1
  creationTimestamp: "2021-04-08T14:33:10Z"
  generateName: emoji-696d9d8f95-
  labels:
    app: emoji-svc
    linkerd.io/control-plane-ns: linkerd
    linkerd.io/proxy-deployment: emoji
    linkerd.io/workload-ns: emojivoto
    pod-template-hash: 696d9d8f95
    version: v11
spec:
  initContainers:
  - args:
    - --incoming-proxy-port
    - "4143"
    - --outgoing-proxy-port
    - "4140"
    - --proxy-uid
    - "2102"
    - --inbound-ports-to-ignore
    - 4190,4191
    - --outbound-ports-to-ignore
    - "44556"
    image: cr.l5d.io/linkerd/proxy-init:v1.3.9
    imagePullPolicy: IfNotPresent
    name: linkerd-init
```
(opaque ports is in there, skip outbound isn't -- although the initContainer gets the right argument since this is already applied from the namespace by the proxy injector).

**After the changes**:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    config.linkerd.io/opaque-ports: "34567"
    config.linkerd.io/proxy-log-level: debug
    config.linkerd.io/skip-outbound-ports: "44556"
    linkerd.io/created-by: linkerd/proxy-injector dev-a7bb62fd-matei
    linkerd.io/identity-mode: default
    linkerd.io/inject: enabled
    linkerd.io/proxy-version: dev-a7bb62fd-matei
  creationTimestamp: "2021-04-08T14:42:06Z"
  generateName: web-5f86686c4d-
  labels:
    app: web-svc
    linkerd.io/control-plane-ns: linkerd
    linkerd.io/proxy-deployment: web
    linkerd.io/workload-ns: emojivoto
    pod-template-hash: 5f86686c4d
    version: v11
  initContainers:
  - args:
    - --incoming-proxy-port
    - "4143"
    - --outgoing-proxy-port
    - "4140"
    - --proxy-uid
    - "2102"
    - --inbound-ports-to-ignore
    - 4190,4191
    - --outbound-ports-to-ignore
    - "44556"
    image: cr.l5d.io/linkerd/proxy-init:v1.3.9
    imagePullPolicy: IfNotPresent
    name: linkerd-init
```
(opaque ports is there and so is skip outbound and the proxy log level, correct options still passed to the initContainers).

*Edit*: made a small change, had a look at `GetNsConfigKeys()` and thought it'd be better to keep the slice of keys as a fixed length array since we know there will be at most `len(ProxyAnnotations)` at any point. Not sure such a big size is warranted but we can avoid calling append for every element.

Signed-off-by: Matei David <matei@buoyant.io>

### CODE REVIEW

File: controller/proxy-injector/webhook.go [hunk 1]
```diff
@@ -88,6 +88,15 @@ func Inject(
 	injectable, reasons := report.Injectable()
 	if injectable {
 		resourceConfig.AppendPodAnnotation(pkgK8s.CreatedByAnnotation, fmt.Sprintf("linkerd/proxy-injector %s", version.Version))
+		// Get all config annotations applied to the namespace
+		// and add them to the pod spec if inexistent
```
Comment (kleimkuhler) [inline controller/proxy-injector/webhook.go:92]: Super nit: if a comment is going to be wrapped then do it at column 80 😬

**Review by alpeb — APPROVED** [review]
Looks great! :+1: 
What do you think about adding support for the `config.alpha.linkerd.io`-prefixed annotations? (currently only used for `/proxy-wait-before-exit-seconds`)

**Review by kleimkuhler — COMMENTED** [review]
Reading through the description the following part stood out to me:

> (opaque ports is in there, skip outbound isn't -- although the initContainer gets the right argument since this is already applied from the namespace by the proxy injector)

Aside from the opaque ports annotation, configuration annotations that are on the namespace are properly copied down to the pod, but do no show up in it's final template. So `skip-outbound-ports` sets the correct environment variables on the pod's init container, but if you look at the pod's template you don't see a `skip-outbound-ports` annotation.

So as this change currently stands, annotations are now added to the pod's template, but nothing has changed about how they are set in the pod's patch. My thought is now that a pod does explicitly inherit annotations from its namespace, it should also change how it gets those annotations values set in the rendered patch.

For example, looking at the injection path of a pod in `webhook.go` we start with `resourceConfig.GetNsConfigKeys()`. This merges the pod's template annotations with the namespace. So, all annotations that we should care about on the namespace are now on the pod.

Next, we `GetPodPatch`, and the first thing that it does is `GetOverriddenValues`. This function looks at the namespace annotations again and merges those values into the resource config's values—but we already have all the namespace annotations.

I think what we should do here is:

1. Have a single location where we look at NS annotations and merge them into the pod's template annotations.
2. Override the resource config's values by looking only at it's own annotations—not the namespace.

**Comment by adleong** [comment]
+1 to @kleimkuhler 's suggestion.  This seems like a good opportunity to simplify this code path a bit.

**Comment by mateiidavid** [comment]
Thank you all for the feedback! Pushed some updates just now.

### Changes

* Wrapped my comments at col. 80 (TIL about `gq` in vim)

* Added support for `config.alpha.linkerd.io-prefixed` annotations as suggested by @alpeb. In doing so, I had two options, either add them to the ProxyAnnotations slice or create a new slice just for "alpha" annotations. I went with the latter since  
adding config annotations with the word alpha in them seems break some functionality; [configToPrometheusLabels()](https://github.com/linkerd/linkerd2/blob/3be5323a00bef7ea0b304758fe9d156a25c8698c/controller/proxy-injector/webhook.go#L76) validates the format of config labels and currently fails if the word alpha is in the annotation key.

* Added `k8s.ProxyLogLevelFormat` to the list of proxy config annotations (twas missing). 

I have also tried to optimize and simplify the injection path. I spent quite a bit of time on this and while my solution doesn't necessarily look simpler, it should be a bit more consistent and eliminate some redundancy. Any suggestions here are more than welcome. Here's what I have done:

* To remove some of the clutter I moved the logic of appending config annotations into its own method -- `conf.AppendNamespaceAnnotations()`. Rather than returning a slice of config keys (what `GetNsConfigKeys()` was doing)  we go through all of the supported annotations; if the namespace has the annotation but the pod doesn't, we add it to the pod.

* I moved the opaque ports check a bit further down in the function -- this would remove a redundant append (in case we're dealing with a pod we'd append the annotation value twice, once when we do the first opaque ports check and another time when we append all values from its namespace). I tested it doesn't break and so far everything works as expected 🤞🏻 my logic here is sound.

* We are no longer using the namespace annotation values as a base when overriding values. Instead, we are using the current pod's annotations as a base and override based on the pod template annotation (i.e what _will_ be applied with the patch). We are only going through the pod template annotation _IFF the injection doesn't happen on the CLI side_. There are two reasons for this:
     1. If we override values through the CLI then the flags will take precedence and will act as the base values before a patch is generated, this means we don't really have to override any values again from the annotation.
     2. If we don't include this check, then the annotation values from the pod template will be parsed differently and will break some tests -- for example, setting overrides on the CLI from the template means opaque ports are "expanded". Let me exemplify: `LINKERD2_PROXY_INBOUND_PORTS_DISABLE_PROTOCOL_DETECTION: 3000-4000` on the CLI, after the change it would be `LINKERD2_PROXY_INBOUND_PORTS_DISABLE_PROTOCOL_DETECTION: 3000, 3001, 3002, ..., 4000"`. 
     
I thought adding a check might be easier to deal with than changing more files (mainly tests).

Think that's about it. Everything is working as before:
```
apiVersion: v1
kind: Namespace
metadata:
  name: emojivoto
  annotations:
    config.linkerd.io/opaque-ports: 34567
    config.linkerd.io/skip-outbound-ports: "44556"
    config.linkerd.io/proxy-log-level: debug
    config.alpha.linkerd.io/proxy-wait-before-exit-seconds: 300
```

```
# appended a star in front of lines where we are using an override/annotation value to make it easier to spot
apiVersion: v1
kind: Pod
metadata:
  annotations:
    config.alpha.linkerd.io/proxy-wait-before-exit-seconds: "300"
    config.linkerd.io/opaque-ports: "34567"
    config.linkerd.io/proxy-log-level: debug
    config.linkerd.io/skip-outbound-ports: "44556"
    linkerd.io/created-by: linkerd/proxy-injector git-3be5323a
    linkerd.io/identity-mode: default
    linkerd.io/inject: enabled
    linkerd.io/proxy-version: git-3be5323a
  creationTimestamp: "2021-04-12T14:40:31Z"
  generateName: voting-ff4c54b8d-
  labels:
    app: voting-svc
    linkerd.io/control-plane-ns: linkerd
    linkerd.io/proxy-deployment: voting
    linkerd.io/workload-ns: emojivoto
    pod-template-hash: ff4c54b8d
    version: v11
pec:
  containers:
 - env:
*    - name: LINKERD2_PROXY_LOG
      value: debug
    - name: LINKERD2_PROXY_LOG_FORMAT
      value: plain
    - name: LINKERD2_PROXY_DESTINATION_SVC_ADDR
      value: linkerd-dst-headless.linkerd.svc.cluster.local.:8086
    - name: LINKERD2_PROXY_DESTINATION_PROFILE_NETWORKS
      value: 10.0.0.0/8,100.64.0.0/10,172.16.0.0/12,192.168.0.0/16
    - name: LINKERD2_PROXY_INBOUND_CONNECT_TIMEOUT
      value: 100ms
    - name: LINKERD2_PROXY_OUTBOUND_CONNECT_TIMEOUT
      value: 1000ms
    - name: LINKERD2_PROXY_CONTROL_LISTEN_ADDR
      value: 0.0.0.0:4190
    - name: LINKERD2_PROXY_ADMIN_LISTEN_ADDR
      value: 0.0.0.0:4191
    - name: LINKERD2_PROXY_OUTBOUND_LISTEN_ADDR
      value: 127.0.0.1:4140
    - name: LINKERD2_PROXY_INBOUND_LISTEN_ADDR
      value: 0.0.0.0:4143
    - name: LINKERD2_PROXY_DESTINATION_PROFILE_SUFFIXES
      value: svc.cluster.local.
    - name: LINKERD2_PROXY_INBOUND_ACCEPT_KEEPALIVE
      value: 10000ms
    - name: LINKERD2_PROXY_OUTBOUND_CONNECT_KEEPALIVE
      value: 10000ms
*   - name: LINKERD2_PROXY_INBOUND_PORTS_DISABLE_PROTOCOL_DETECTION
      value: "34567"
    ...
    image: cr.l5d.io/linkerd/proxy:git-3be5323a
    imagePullPolicy: IfNotPresent
    lifecycle:
      preStop:
        exec:
          command:
          - /bin/bash
          - -c
 *         - sleep 300
  initContainers:
  - args:
    - --incoming-proxy-port
    - "4143"
    - --outgoing-proxy-port
    - "4140"
    - --proxy-uid
    - "2102"
    - --inbound-ports-to-ignore
    - 4190,4191
    - --outbound-ports-to-ignore
 *   - "44556"
    image: cr.l5d.io/linkerd/proxy-init:v1.3.9
    imagePullPolicy: IfNotPresent
    name: linkerd-init

```

*TL;DR*: addressed all comments, refactored function to get config keys into a method, should loop less now. No longer using namespace annotations as base but to use the pod template's annotations in overrides implies the use of an origin check so that opaque ports ranges don't get expanded.

**Review by adleong — APPROVED** [review]
Thanks for updating, @mateiidavid!  While the code is still a bit complex, I think that's an accurate reflection of the complexity of the logic that we're implementing here.  One thing that's worth noting is that if a pod is not injectable, the opaque-ports annotation will be copied to that pod from the namespace annotations, but other config annotations will not.  This is perhaps a bit surprising but I believe it's probably the behavior that we want. :+1:

**Review by kleimkuhler — APPROVED** [review]
Yep this lines up with what I expected after the first review! Good catch on adding `ProxyLogFormatAnnotation` 👍
